daft/runners/ray_runner.py

```diff
@@ -711,136 +849,65 @@ def place_in_queue(item):
-                                                boundaries=single_partial.boundaries,
-                                            )
-                                        ]
-                                    )
-
-                                next_step.set_result(
-                                    [RayMaterializedResult(partition, accessor, 0) for partition in next_step.inputs]
-                                )
-                                next_step.set_done()
-                                next_step = next(tasks)
-
-                            else:
-                                # Add the task to the batch.
-                                tasks_to_dispatch.append(next_step)
-                                next_step = next(tasks)
-
-                        # Dispatch the batch of tasks.
+                        # Dispatch Batching
+                        tasks_to_dispatch, has_next = self._construct_dispatch_batch(
+                            result_uuid,
+                            tasks,
+                            dispatches_allowed,
+                        )
+
                         logger.debug(
                             "%ss: RayRunner dispatching %s tasks",
                             (datetime.now() - start).total_seconds(),
                             len(tasks_to_dispatch),
                         )
 
-                        if not is_active():
+                        if not self._is_active(result_uuid):
                             break
 
-                        for task in tasks_to_dispatch:
-                            if task.actor_pool_id is None:
-                                results = _build_partitions(daft_execution_config, task)
-                            else:
-                                actor_pool = self._actor_pools.get(task.actor_pool_id)
-                                assert actor_pool is not None, "Ray actor pool must live for as long as the tasks."
-                                results = _build_partitions_on_actor_pool(task, actor_pool)
-                            logger.debug("%s -> %s", task, results)
+                        # Dispatch
+                        for task, result_obj_refs in self._dispatch_tasks(
+                            tasks_to_dispatch,
+                            daft_execution_config,
+                        ):
                             inflight_tasks[task.id()] = task
-                            for result in results:
+                            for result in result_obj_refs:
                                 inflight_ref_to_task[result] = task.id()
 
                             pbar.mark_task_start(task)
 
-                        if dispatches_allowed == 0 or next_step is None:
+                        # Break the dispatch batching/dispatch loop if no more dispatches allowed, or physical plan
+                        # needs work for forward progress
+                        if dispatches_allowed == 0 or not has_next:
```

Comment:
NOTE: behavior change here. We now use a has_next variable that is returned from self._construct_dispatch_batch to figure out whether or now we should break the DispatchBatching -> Dispatch loop.

Previously the scheduling loop had to keep track of next_step which was super weird and unwieldy, and caused us to call next(tasks) in a bunch of random places scattered throughout the loop.